File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.java
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode slow = head;
        ListNode fast = head;

        // Move fast n steps ahead
        for (int i = 1; i <= n; i++) {
            fast = fast.next;
        }

        // If fast is null, it means we need to remove the head (n == list length)
        if (fast == null) {
            return head.next; // Remove the head by returning head.next
        }

        // Move both fast and slow pointers until fast reaches the end
        while (fast.next != null) {
            slow = slow.next;
            fast = fast.next;
        }

        // Now slow is at the node before the one to remove
        slow.next = slow.next.next;

        // Return the original head (not slow)
        return head;
    }
}

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def lengthofList(self, head: ListNode) -> int:
        size = 0
        while head is not None:
            size += 1
            head = head.next
        return size

    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        if head is None:
            return head

        sizeOfList = self.lengthofList(head)
        dltNode = sizeOfList - n

        # If we need to delete the head node
        if dltNode == 0:
            return head.next

        pointer = head
        for i in range(dltNode - 1):  # go to node just before the one to delete
            pointer = pointer.next

        # delete the nth node
        if pointer.next is not None:
            pointer.next = pointer.next.next

        return head
